plugin/discovery/get.go

```diff
@@ -111,7 +112,9 @@ type ProviderInstaller struct {
 // are produced under the assumption that if presented to the user they will
 // be presented alongside context about what is being installed, and thus the
 // error messages do not redundantly include such information.
-func (i *ProviderInstaller) Get(provider string, req Constraints) (PluginMeta, error) {
+func (i *ProviderInstaller) Get(provider string, req Constraints) (PluginMeta, tfdiags.Diagnostics, error) {
```

Comment:
It's a little strange to see both diagnostics _and_ error in the return values here since normally we use diagnostics _in place of_ `error`, but I can see why you did it and I think we can live with it as an interim step on the path towards a more general refactor of how we split the concerns between UI and installer here, which can come at some later point.

Note that there's another caller of this in our `terraform-bundle` helper tool:

https://github.com/hashicorp/terraform/blob/2b726a892f809fa4a9cf95b1020f5f54c092761c/tools/terraform-bundle/package.go#L184

That tool will eventually be rendered redundant by generalized provider installation from the registry, but we need to keep it working for the moment so that PTFE users (and other situations where Terraform CLI can't access the Internet to install on the fly) can easily package up Terraform with a selection of providers for local installation.

Reply:
Agreed on diags + error returned but yes, changing all of the errors returned seemed like a larger refactoring.

Thanks for catching that! Fixed by ignoring with `_`, as I didn't see a showDiagnostics method available in that command's scope, but I'm happy to copy it to that package if you'd life.

Reply:
Indeed that separate tool predates the idea of diagnostics and so doesn't have the infrastructure in place to return them properly. Given that the remaining life of this tool is planned to be short and right now the only thing returned via the diagnostics channel is warnings I think ignoring them is okay... the implication there is that if the registry returns warnings in _that_ case then they won't be seen, but that seems livable when our end-game here is for alternate registries to remove the need for this tool altogether.